File: cli/go/awid/stable_identity_forgery_repro_test.go
package awid

import (
	"crypto/ed25519"
	"testing"
	"time"
)

// TestForgedRotationSpoofsIdentity_Repro demonstrates the aajc.3 vulnerability:
// the DID-log verifier accepts a seq>1 rotation whose authorized_by /
// previous_did_key are keys the attacker fully controls, never binding them to
// the victim's genesis key. A registry-confirmed OK_VERIFIED for a victim's
// did:aw can therefore be produced entirely from attacker-held keys.
//
// After the fix this MUST NOT return OK_VERIFIED: an unanchored seq>1 head can
// only ever be OK_DEGRADED (forcing full-log verification from genesis).
func TestForgedRotationSpoofsIdentity_Repro(t *testing.T) {
	t.Parallel()

	// The victim's genuine genesis key -> the did:aw the attacker wants to steal.
	victimPub, _, err := ed25519.GenerateKey(nil)
	if err != nil {
		t.Fatal(err)
	}
	victimStableID := ComputeStableID(victimPub)

	// Attacker keys, unrelated to the victim.
	attackerPrevPub, attackerPrevPriv, err := ed25519.GenerateKey(nil)
	if err != nil {
		t.Fatal(err)
	}
	attackerNewPub, _, err := ed25519.GenerateKey(nil)
	if err != nil {
		t.Fatal(err)
	}
	attackerPrevDID := ComputeDIDKey(attackerPrevPub)
	attackerNewDID := ComputeDIDKey(attackerNewPub)
	prevEntryHash := "0000000000000000000000000000000000000000000000000000000000000000"

	// A self-consistent seq-2 "rotation" signed by the attacker's own key,
	// claiming to rotate the victim's did:aw to an attacker-held key.
	forged := signedDidKeyResolution(t, attackerPrevPriv, &DidKeyResolution{
		DIDAW:         victimStableID,
		CurrentDIDKey: attackerNewDID,
		LogHead: &DidKeyEvidence{
			Seq:            2,
			Operation:      "rotate_key",
			PreviousDIDKey: &attackerPrevDID,
			NewDIDKey:      attackerNewDID,
			PrevEntryHash:  &prevEntryHash,
			StateHash:      stableIdentityStateHash(victimStableID, attackerNewDID),
			AuthorizedBy:   attackerPrevDID,
			Timestamp:      "2026-02-22T10:05:00Z",
		},
	})

	outcome, head, err := VerifyDidKeyResolution(forged, nil, time.Unix(0, 0))

	if outcome == StableIdentityVerified {
		t.Fatalf("VULNERABLE: forged rotation accepted as OK_VERIFIED (head=%+v, err=%v)", head, err)
	}
	if outcome != StableIdentityDegraded {
		t.Fatalf("unanchored seq>1 head must be OK_DEGRADED, got %q (err=%v)", outcome, err)
	}
}

File: channel-core/test/stable_identity_forgery_repro.test.ts
import { describe, expect, test } from "vitest";
import { sha256, sha512 } from "@noble/hashes/sha2.js";
import * as ed from "@noble/ed25519";
import { computeDIDKey, computeStableID } from "../src/identity/did.js";
import {
  canonicalDidLogPayload,
  verifyDidKeyResolution,
  type DidKeyResolution,
} from "../src/identity/registry.js";

ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));

function seed(byte: number): Uint8Array {
  return new Uint8Array(32).fill(byte);
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

// aajc.3 repro: the TS DID-log verifier accepts a seq>1 rotation whose
// authorized_by / previous_did_key are attacker-held keys, never bound to the
// victim's genesis key. After the fix an unanchored seq>1 head MUST be
// OK_DEGRADED, never OK_VERIFIED.
describe("forged rotation spoofing (aajc.3 repro)", () => {
  test("forged rotation must not be accepted as OK_VERIFIED", () => {
    // Victim genesis key -> the did:aw the attacker wants to steal.
    const victimPub = ed.getPublicKey(seed(0x11));
    const victimStableID = computeStableID(victimPub);

    // Attacker keys, unrelated to the victim.
    const attackerPrevPriv = seed(0x22);
    const attackerPrevPub = ed.getPublicKey(attackerPrevPriv);
    const attackerNewPub = ed.getPublicKey(seed(0x33));
    const attackerPrevDID = computeDIDKey(attackerPrevPub);
    const attackerNewDID = computeDIDKey(attackerNewPub);

    const head = {
      authorized_by: attackerPrevDID,
      new_did_key: attackerNewDID,
      operation: "rotate_key",
      prev_entry_hash: "0".repeat(64),
      previous_did_key: attackerPrevDID,
      seq: 2,
      // Canonical state hash for the attacker's replacement key.
      state_hash: bytesToHex(
        sha256(
          new TextEncoder().encode(
            `{"current_did_key":"${attackerNewDID}","did_aw":"${victimStableID}"}`,
          ),
        ),
      ),
      timestamp: "2026-02-22T10:05:00Z",
    };

    const payload = canonicalDidLogPayload(victimStableID, {
      ...head,
      entry_hash: "",
      signature: "",
    });
    const entryHash = bytesToHex(sha256(new TextEncoder().encode(payload)));
    // Unpadded standard base64, matching Go's RawStdEncoding — the wire contract.
    // A padded signature is rejected by both runtimes (default-aajc.8).
    const signature = Buffer.from(
      ed.sign(new TextEncoder().encode(payload), attackerPrevPriv),
    ).toString("base64url").replace(/-/g, "+").replace(/_/g, "/");

    const resolution: DidKeyResolution = {
      did_aw: victimStableID,
      current_did_key: attackerNewDID,
      log_head: { ...head, entry_hash: entryHash, signature },
    };

    const result = verifyDidKeyResolution(resolution, undefined, Date.now());

    expect(result.outcome).not.toBe("OK_VERIFIED");
    expect(result.outcome).toBe("OK_DEGRADED");
  });
});
